cpp/src/Ice/ConnectionI.cpp

```diff
@@ -2512,6 +2534,41 @@ Ice::ConnectionI::initiateShutdown()
     }
 }
 
+void
+Ice::ConnectionI::idleCheck(
+    const IceUtil::TimerTaskPtr& idleCheckTimerTask,
+    const chrono::milliseconds& idleTimeout) noexcept
+{
+    std::lock_guard lock(_mutex);
+    if (_state == StateActive)
+    {
+        // _timer->cancel(task) returns true if a concurrent read rescheduled the timer task.
+        if (_transceiver->isWaitingToBeRead() || _timer->cancel(idleCheckTimerTask))
+        {
+            // Schedule or reschedule timer task. Reschedule in the rare case where a concurrent read scheduled the task
+            // already.
+            _timer->schedule(idleCheckTimerTask, idleTimeout, true);
```

Comment:
Typically, when this function runs, we're going to abort to connection. 
The question is when are we forgiving and don't abort it.

Case 1: the connection is no longer active. In this case, we don't do anything.

Case 2: the connection is waiting to be read, is has not been read yet because we're out of thread. In this case, we reschedule the idleCheck in idle timeout

Case 3: as this idle check is running, another thread is concurrently reading the transceiver and has just (re)scheduled this timer task. The only we can find out of this happened is by canceling the idle timer task (with cancel). When timer->cancel(task) returns true, it means we're in this situation.

And since we've removed the timer task to check if it was there, we need to add it back.

The `if` simply checks "case 2" and "case 3" together, and the action is the same: reschedule timer task.

Other cases: we abort the connection

Reply:
> It is not clear to me under what conditions we get "no data available to be read", and cancel returns false

That's the common abort case. The peer didn't send any data for idle time and the idle check has not been rescheduled by a concurrent read.

Keep in mind that timer _removes_ the timer task from its maps just before it executes it. See the comment on TimerTask::cancel.